src/service/viewer-impl.js

```diff
@@ -766,16 +759,15 @@ export class Viewer {
     if (eventType == 'viewport') {
       if (data['paddingTop'] !== undefined) {
         this.paddingTop_ = data['paddingTop'];
-        this.viewportObservable_.fire(
-          /** @type {!JSONType} */ (data));
+        this.messageObservables_[eventType].fire(
+            /** @type {!JSONType} */ (data));
         return Promise.resolve();
       }
       return undefined;
     }
     if (eventType == 'historyPopped') {
-      this.historyPoppedObservable_.fire({
-        newStackIndex: data['newStackIndex'],
-      });
+      this.messageObservables_[eventType].fire(
```

Comment:
We could instead consider allowing handlers to return values too. Let me know if you like that option better. But I thought `ViewerMessage` would work well.